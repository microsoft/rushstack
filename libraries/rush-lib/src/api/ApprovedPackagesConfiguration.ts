// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'node:path';

import { JsonFile, JsonSchema, FileSystem, NewlineKind, InternalError } from '@rushstack/node-core-library';

import { JsonSchemaUrls } from '../logic/JsonSchemaUrls';
import schemaJson from '../schemas/approved-packages.schema.json';
import { RushConstants } from '../logic/RushConstants';

/**
 * Part of IApprovedPackagesJson.
 */
export interface IApprovedPackagesItemJson {
  name: string;
  allowedCategories: string[];
}

/**
 * This represents the JSON data structure for the "browser-approved-packages.json"
 * and "nonbrowser-approved-packages.json" configuration files.  See "approved-packages.schema.json"
 * for documentation.
 */
export interface IApprovedPackagesJson {
  $schema?: string;
  packages: IApprovedPackagesItemJson[];
}

/**
 * An item returned by ApprovedPackagesConfiguration
 * @public
 */
export class ApprovedPackagesItem {
  /**
   * The NPM package name
   */
  public packageName: string;

  /**
   * The project categories that are allowed to use this package.
   */
  public allowedCategories: Set<string> = new Set<string>();

  /**
   * @internal
   */
  public constructor(packageName: string) {
    this.packageName = packageName;
  }
}

/**
 * This represents the JSON file specified via the "approvedPackagesFile" option in rush.json.
 * @public
 */
export class ApprovedPackagesConfiguration {
  private static _jsonSchema: JsonSchema = JsonSchema.fromLoadedObject(schemaJson);

  public items: ApprovedPackagesItem[] = [];

  private _itemsByName: Map<string, ApprovedPackagesItem> = new Map<string, ApprovedPackagesItem>();

  private _loadedJson!: IApprovedPackagesJson;
  private _jsonFilename: string;

  public constructor(jsonFilename: string) {
    this._jsonFilename = jsonFilename;
    this.clear();
  }

  /**
   * Clears all the settings, returning to an empty state.
   */
  public clear(): void {
    this._itemsByName.clear();
    this._loadedJson = {
      // Ensure this comes first in the key ordering
      $schema: '',
      packages: []
    };
  }

  public getItemByName(packageName: string): ApprovedPackagesItem | undefined {
    return this._itemsByName.get(packageName);
  }

  public addOrUpdatePackage(packageName: string, reviewCategory: string): boolean {
    let changed: boolean = false;

    let item: ApprovedPackagesItem | undefined = this._itemsByName.get(packageName);
    if (!item) {
      item = new ApprovedPackagesItem(packageName);
      this._addItem(item);
      changed = true;
    }

    if (reviewCategory && !item.allowedCategories.has(reviewCategory)) {
      item.allowedCategories.add(reviewCategory);
      changed = true;
    }

    return changed;
  }

  /**
   * If the file exists, calls loadFromFile().
   */
  public tryLoadFromFile(approvedPackagesPolicyEnabled: boolean): boolean {
    if (!FileSystem.exists(this._jsonFilename)) {
      return false;
    }

    this.loadFromFile();

    if (!approvedPackagesPolicyEnabled) {
      // eslint-disable-next-line no-console
      console.log(
        `Warning: Ignoring "${path.basename(this._jsonFilename)}" because the` +
          ` "approvedPackagesPolicy" setting was not specified in ${RushConstants.rushJsonFilename}`
      );
    }

    return false;
  }

  /**
   * Loads the configuration data from the filename that was passed to the constructor.
   */
  public loadFromFile(): void {
    const approvedPackagesJson: IApprovedPackagesJson = JsonFile.loadAndValidate(
      this._jsonFilename,
      ApprovedPackagesConfiguration._jsonSchema
    );

    this.clear();

    for (const browserPackage of approvedPackagesJson.packages) {
      this._addItemJson(browserPackage, this._jsonFilename);
    }
  }

  /**
   * Loads the configuration data to the filename that was passed to the constructor.
   */
  public saveToFile(): void {
    // Update the JSON structure that we already loaded, preserving any existing state
    // (which passed schema validation).

    // eslint-disable-next-line dot-notation
    this._loadedJson['$schema'] = JsonSchemaUrls.approvedPackages;

    this._loadedJson.packages = [];

    this.items.sort((a: ApprovedPackagesItem, b: ApprovedPackagesItem) => {
      return a.packageName.localeCompare(b.packageName);
    });

    for (const item of this.items) {
      // Sort the items from the set.
      const allowedCategories: string[] = Array.from(item.allowedCategories);
      allowedCategories.sort();

      const itemJson: IApprovedPackagesItemJson = {
        name: item.packageName,
        allowedCategories: allowedCategories
      };

      this._loadedJson.packages.push(itemJson);
    }

    // Save the file
    let body: string = JsonFile.stringify(this._loadedJson);

    // Unindent the allowedCategories array to improve readability
    body = body.replace(/("allowedCategories": +\[)([^\]]+)/g, (substring: string, ...args: string[]) => {
      return args[0] + args[1].replace(/\s+/g, ' ');
    });

    // Add a header
    body = '// DO NOT ADD COMMENTS IN THIS FILE.  They will be lost when the Rush tool resaves it.\n' + body;

    FileSystem.writeFile(this._jsonFilename, body, {
      convertLineEndings: NewlineKind.CrLf
    });
  }

  /**
   * Helper function only used by the constructor when loading the file.
   */
  private _addItemJson(itemJson: IApprovedPackagesItemJson, jsonFilename: string): void {
    if (this._itemsByName.has(itemJson.name)) {
      throw new Error(
        `Error loading package review file ${jsonFilename}:\n` +
          ` the name "${itemJson.name}" appears more than once`
      );
    }

    const item: ApprovedPackagesItem = new ApprovedPackagesItem(itemJson.name);
    if (itemJson.allowedCategories) {
      for (const allowedCategory of itemJson.allowedCategories) {
        item.allowedCategories.add(allowedCategory);
      }
    }
    this._addItem(item);
  }

  /**
   * Helper function that adds an already created ApprovedPackagesItem to the
   * list and set.
   */
  private _addItem(item: ApprovedPackagesItem): void {
    if (this._itemsByName.has(item.packageName)) {
      throw new InternalError('Duplicate key');
    }
    this.items.push(item);
    this._itemsByName.set(item.packageName, item);
  }
}
