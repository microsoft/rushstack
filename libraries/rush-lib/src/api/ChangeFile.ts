// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'node:path';

import type gitInfo from 'git-repo-info';

import { JsonFile } from '@rushstack/node-core-library';

import type { RushConfiguration } from './RushConfiguration';
import type { IChangeFile, IChangeInfo } from './ChangeManagement';
import { Git } from '../logic/Git';

/**
 * This class represents a single change file.
 */
export class ChangeFile {
  private _changeFileData: IChangeFile;
  private _rushConfiguration: RushConfiguration;

  /**
   * @internal
   */
  public constructor(changeFileData: IChangeFile, rushConfiguration: RushConfiguration) {
    if (!changeFileData) {
      throw new Error(`changeFileData does not have a value`);
    }

    if (!rushConfiguration) {
      throw new Error(`rushConfiguration does not have a value`);
    }

    this._changeFileData = changeFileData;
    this._rushConfiguration = rushConfiguration;
  }

  /**
   * Adds a change entry into the change file
   * @param data - change information
   */
  public addChange(data: IChangeInfo): void {
    this._changeFileData.changes.push(data);
  }

  /**
   * Gets all the change entries about the specified package from the change file.
   * @param packageName - package name
   */
  public getChanges(packageName: string): IChangeInfo[] {
    const changes: IChangeInfo[] = [];
    for (const info of this._changeFileData.changes) {
      if (info.packageName === packageName) {
        changes.push(info);
      }
    }
    return changes;
  }

  /**
   * Writes the change file to disk in sync mode.
   * Returns the file path.
   * @returns the path to the file that was written (based on generatePath())
   */
  public writeSync(): string {
    const filePath: string = this.generatePath();
    JsonFile.save(this._changeFileData, filePath, {
      ensureFolderExists: true
    });
    return filePath;
  }

  /**
   * Generates a file path for storing the change file to disk.
   * Note that this value may change if called twice in a row,
   * as it is partially based on the current date/time.
   */
  public generatePath(): string {
    let branch: string | undefined = undefined;
    const git: Git = new Git(this._rushConfiguration);
    const repoInfo: gitInfo.GitRepoInfo | undefined = git.getGitInfo();
    branch = repoInfo && repoInfo.branch;
    if (!branch) {
      // eslint-disable-next-line no-console
      console.log('Could not automatically detect git branch name, using timestamp instead.');
    }

    // example filename: yourbranchname_2017-05-01-20-20.json
    const filename: string = branch
      ? this._escapeFilename(`${branch}_${this._getTimestamp()}.json`)
      : `${this._getTimestamp()}.json`;
    const filePath: string = path.join(
      this._rushConfiguration.changesFolder,
      ...this._changeFileData.packageName.split('/'),
      filename
    );
    return filePath;
  }

  /**
   * Gets the current time, formatted as YYYY-MM-DD-HH-MM
   * Optionally will include seconds
   */
  private _getTimestamp(useSeconds: boolean = false): string | undefined {
    // Create a date string with the current time

    // dateString === "2016-10-19T22:47:49.606Z"
    const dateString: string = new Date().toJSON();

    // Parse out 2 capture groups, the date and the time
    const dateParseRegex: RegExp = /([0-9]{4}-[0-9]{2}-[0-9]{2}).*([0-9]{2}:[0-9]{2}:[0-9]{2})/;

    // matches[1] === "2016-10-19"
    // matches[2] === "22:47:49"
    const matches: RegExpMatchArray | null = dateString.match(dateParseRegex);

    if (matches) {
      // formattedDate === "2016-10-19"
      const formattedDate: string = matches[1];

      let formattedTime: string;
      if (useSeconds) {
        // formattedTime === "22-47-49"
        formattedTime = matches[2].replace(':', '-');
      } else {
        // formattedTime === "22-47"
        const timeParts: string[] = matches[2].split(':');
        formattedTime = `${timeParts[0]}-${timeParts[1]}`;
      }

      return `${formattedDate}-${formattedTime}`;
    }
    return undefined;
  }

  private _escapeFilename(filename: string, replacer: string = '-'): string {
    // Removes / ? < > \ : * | ", really anything that isn't a letter, number, '.' '_' or '-'
    const badCharacters: RegExp = /[^a-zA-Z0-9._-]/g;
    return filename.replace(badCharacters, replacer);
  }
}
