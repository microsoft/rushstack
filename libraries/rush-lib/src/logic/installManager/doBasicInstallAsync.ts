// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { ITerminal } from '@rushstack/terminal';

import type { RushConfiguration } from '../../api/RushConfiguration';
import type { RushGlobalFolder } from '../../api/RushGlobalFolder';
import type { BaseInstallManager } from '../base/BaseInstallManager';
import type { IInstallManagerOptions } from '../base/BaseInstallManagerTypes';
import { InstallManagerFactory } from '../InstallManagerFactory';
import { SetupChecks } from '../SetupChecks';
import { PurgeManager } from '../PurgeManager';
import { VersionMismatchFinder } from '../versionMismatch/VersionMismatchFinder';
import type { Subspace } from '../../api/Subspace';

export interface IRunInstallOptions {
  afterInstallAsync?: IInstallManagerOptions['afterInstallAsync'];
  beforeInstallAsync?: IInstallManagerOptions['beforeInstallAsync'];
  rushConfiguration: RushConfiguration;
  rushGlobalFolder: RushGlobalFolder;
  isDebug: boolean;
  terminal: ITerminal;
  variant: string | undefined;
  subspace: Subspace;
}

export async function doBasicInstallAsync(options: IRunInstallOptions): Promise<void> {
  const {
    rushConfiguration,
    rushGlobalFolder,
    isDebug,
    variant,
    terminal,
    beforeInstallAsync,
    afterInstallAsync,
    subspace
  } = options;

  VersionMismatchFinder.ensureConsistentVersions(rushConfiguration, terminal, {
    variant,
    subspace
  });
  SetupChecks.validate(rushConfiguration);

  const purgeManager: typeof PurgeManager.prototype = new PurgeManager(rushConfiguration, rushGlobalFolder);

  const installManager: BaseInstallManager = await InstallManagerFactory.getInstallManagerAsync(
    rushConfiguration,
    rushGlobalFolder,
    purgeManager,
    {
      debug: isDebug,
      allowShrinkwrapUpdates: false,
      checkOnly: false,
      bypassPolicy: false,
      noLink: false,
      fullUpgrade: false,
      recheckShrinkwrap: false,
      offline: false,
      collectLogFile: false,
      pnpmFilterArgumentValues: [],
      selectedProjects: new Set(rushConfiguration.projects),
      maxInstallAttempts: 1,
      networkConcurrency: undefined,
      subspace,
      terminal,
      variant,
      afterInstallAsync,
      beforeInstallAsync
    }
  );

  try {
    await installManager.doInstallAsync();
  } finally {
    await purgeManager.startDeleteAllAsync();
  }
}
