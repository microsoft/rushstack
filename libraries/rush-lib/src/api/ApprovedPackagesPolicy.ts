// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'node:path';

import { ApprovedPackagesConfiguration } from './ApprovedPackagesConfiguration';
import { RushConstants } from '../logic/RushConstants';
import type {
  RushConfiguration,
  IRushConfigurationJson,
  IApprovedPackagesPolicyJson
} from './RushConfiguration';

/**
 * This is a helper object for RushConfiguration.
 * It exposes the "approvedPackagesPolicy" feature from rush.json.
 * @public
 */
export class ApprovedPackagesPolicy {
  /**
   * Whether the feature is enabled.  The feature is enabled if the "approvedPackagesPolicy"
   * field is assigned in rush.json.
   */
  public readonly enabled: boolean;

  /**
   * A list of NPM package scopes that will be excluded from review (e.g. `@types`)
   */
  public readonly ignoredNpmScopes: ReadonlySet<string>;

  /**
   * A list of category names that are valid for usage as the RushConfigurationProject.reviewCategory field.
   * This array will never be undefined.
   */
  public readonly reviewCategories: ReadonlySet<string>;

  /**
   * Packages approved for usage in a web browser.  This is the stricter of the two types, so by default
   * all new packages are added to this file.
   *
   * @remarks
   *
   * This is part of an optional approval workflow, whose purpose is to review any new dependencies
   * that are introduced (e.g. maybe a legal review is required, or maybe we are trying to minimize bloat).
   * When Rush discovers a new dependency has been added to package.json, it will update the file.
   * The intent is that the file will be stored in Git and tracked by a branch policy that notifies
   * reviewers when a PR attempts to modify the file.
   *
   * Example filename: `C:\MyRepo\common\config\rush\browser-approved-packages.json`
   */
  public readonly browserApprovedPackages: ApprovedPackagesConfiguration;

  /**
   * Packages approved for usage everywhere *except* in a web browser.
   *
   * @remarks
   *
   * This is part of an optional approval workflow, whose purpose is to review any new dependencies
   * that are introduced (e.g. maybe a legal review is required, or maybe we are trying to minimize bloat).
   * The intent is that the file will be stored in Git and tracked by a branch policy that notifies
   * reviewers when a PR attempts to modify the file.
   *
   * Example filename: `C:\MyRepo\common\config\rush\browser-approved-packages.json`
   */
  public readonly nonbrowserApprovedPackages: ApprovedPackagesConfiguration;

  /** @internal */
  public constructor(rushConfiguration: RushConfiguration, rushConfigurationJson: IRushConfigurationJson) {
    const approvedPackagesPolicy: IApprovedPackagesPolicyJson =
      rushConfigurationJson.approvedPackagesPolicy || {};

    this.enabled = !!rushConfigurationJson.approvedPackagesPolicy;
    this.ignoredNpmScopes = new Set<string>(approvedPackagesPolicy.ignoredNpmScopes);
    this.reviewCategories = new Set<string>(approvedPackagesPolicy.reviewCategories);

    if (this.enabled) {
      if (!this.reviewCategories.size) {
        throw new Error(
          `The "approvedPackagesPolicy" feature is enabled ${RushConstants.rushJsonFilename}, but the reviewCategories` +
            ` list is not configured.`
        );
      }
    }

    // Load browser-approved-packages.json
    const browserApprovedPackagesPath: string = path.join(
      rushConfiguration.commonRushConfigFolder,
      RushConstants.browserApprovedPackagesFilename
    );
    this.browserApprovedPackages = new ApprovedPackagesConfiguration(browserApprovedPackagesPath);
    this.browserApprovedPackages.tryLoadFromFile(this.enabled);

    // Load nonbrowser-approved-packages.json
    const nonbrowserApprovedPackagesPath: string = path.join(
      rushConfiguration.commonRushConfigFolder,
      RushConstants.nonbrowserApprovedPackagesFilename
    );
    this.nonbrowserApprovedPackages = new ApprovedPackagesConfiguration(nonbrowserApprovedPackagesPath);
    this.nonbrowserApprovedPackages.tryLoadFromFile(this.enabled);
  }
}
