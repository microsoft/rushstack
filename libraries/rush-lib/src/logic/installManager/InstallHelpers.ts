// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'path';
import {
  FileConstants,
  FileSystem,
  type IPackageJson,
  JsonFile,
  LockFile
} from '@rushstack/node-core-library';
import { Colorize, type ITerminal } from '@rushstack/terminal';

import { LastInstallFlag } from '../../api/LastInstallFlag';
import type { PackageManagerName } from '../../api/packageManager/PackageManager';
import type { RushConfiguration } from '../../api/RushConfiguration';
import type { RushGlobalFolder } from '../../api/RushGlobalFolder';
import { Utilities } from '../../utilities/Utilities';
import type { IConfigurationEnvironment } from '../base/BasePackageManagerOptionsConfiguration';
import type { PnpmOptionsConfiguration } from '../pnpm/PnpmOptionsConfiguration';
import { merge } from '../../utilities/objectUtilities';
import type { Subspace } from '../../api/Subspace';
import { RushConstants } from '../RushConstants';
import * as semver from 'semver';

interface ICommonPackageJson extends IPackageJson {
  pnpm?: {
    overrides?: typeof PnpmOptionsConfiguration.prototype.globalOverrides;
    packageExtensions?: typeof PnpmOptionsConfiguration.prototype.globalPackageExtensions;
    peerDependencyRules?: typeof PnpmOptionsConfiguration.prototype.globalPeerDependencyRules;
    neverBuiltDependencies?: typeof PnpmOptionsConfiguration.prototype.globalNeverBuiltDependencies;
    ignoredOptionalDependencies?: typeof PnpmOptionsConfiguration.prototype.globalIgnoredOptionalDependencies;
    allowedDeprecatedVersions?: typeof PnpmOptionsConfiguration.prototype.globalAllowedDeprecatedVersions;
    patchedDependencies?: typeof PnpmOptionsConfiguration.prototype.globalPatchedDependencies;
  };
}

export class InstallHelpers {
  public static generateCommonPackageJson(
    rushConfiguration: RushConfiguration,
    subspace: Subspace,
    dependencies: Map<string, string> = new Map<string, string>(),
    terminal: ITerminal
  ): void {
    const commonPackageJson: ICommonPackageJson = {
      dependencies: {},
      description: 'Temporary file generated by the Rush tool',
      name: 'rush-common',
      private: true,
      version: '0.0.0'
    };

    if (rushConfiguration.isPnpm) {
      const pnpmOptions: PnpmOptionsConfiguration =
        subspace.getPnpmOptions() || rushConfiguration.pnpmOptions;
      if (!commonPackageJson.pnpm) {
        commonPackageJson.pnpm = {};
      }

      if (pnpmOptions.globalOverrides) {
        commonPackageJson.pnpm.overrides = pnpmOptions.globalOverrides;
      }

      if (pnpmOptions.globalPackageExtensions) {
        commonPackageJson.pnpm.packageExtensions = pnpmOptions.globalPackageExtensions;
      }
      if (pnpmOptions.globalPeerDependencyRules) {
        commonPackageJson.pnpm.peerDependencyRules = pnpmOptions.globalPeerDependencyRules;
      }

      if (pnpmOptions.globalNeverBuiltDependencies) {
        commonPackageJson.pnpm.neverBuiltDependencies = pnpmOptions.globalNeverBuiltDependencies;
      }

      if (pnpmOptions.globalIgnoredOptionalDependencies) {
        if (
          rushConfiguration.rushConfigurationJson.pnpmVersion !== undefined &&
          semver.lt(rushConfiguration.rushConfigurationJson.pnpmVersion, '9.0.0')
        ) {
          terminal.writeWarningLine(
            Colorize.yellow(
              `Your version of pnpm (${rushConfiguration.rushConfigurationJson.pnpmVersion}) ` +
                `doesn't support the "globalIgnoredOptionalDependencies" field in ` +
                `${rushConfiguration.commonRushConfigFolder}/${RushConstants.pnpmConfigFilename}. ` +
                'Remove this field or upgrade to pnpm 9.'
            )
          );
        }

        commonPackageJson.pnpm.ignoredOptionalDependencies = pnpmOptions.globalIgnoredOptionalDependencies;
      }

      if (pnpmOptions.globalAllowedDeprecatedVersions) {
        commonPackageJson.pnpm.allowedDeprecatedVersions = pnpmOptions.globalAllowedDeprecatedVersions;
      }

      if (pnpmOptions.globalPatchedDependencies) {
        commonPackageJson.pnpm.patchedDependencies = pnpmOptions.globalPatchedDependencies;
      }

      if (pnpmOptions.unsupportedPackageJsonSettings) {
        merge(commonPackageJson, pnpmOptions.unsupportedPackageJsonSettings);
      }
    }

    // Add any preferred versions to the top of the commonPackageJson
    // do this in alphabetical order for simpler debugging
    for (const dependency of Array.from(dependencies.keys()).sort()) {
      commonPackageJson.dependencies![dependency] = dependencies.get(dependency)!;
    }

    // Example: "C:\MyRepo\common\temp\package.json"
    const commonPackageJsonFilename: string = path.join(
      subspace.getSubspaceTempFolderPath(),
      FileConstants.PackageJson
    );

    // Don't update the file timestamp unless the content has changed, since "rush install"
    // will consider this timestamp
    JsonFile.save(commonPackageJson, commonPackageJsonFilename, { onlyIfChanged: true });
  }

  public static getPackageManagerEnvironment(
    rushConfiguration: RushConfiguration,
    options: {
      debug?: boolean;
    } = {}
  ): NodeJS.ProcessEnv {
    let configurationEnvironment: IConfigurationEnvironment | undefined = undefined;

    if (rushConfiguration.packageManager === 'npm') {
      if (rushConfiguration.npmOptions && rushConfiguration.npmOptions.environmentVariables) {
        configurationEnvironment = rushConfiguration.npmOptions.environmentVariables;
      }
    } else if (rushConfiguration.isPnpm) {
      if (rushConfiguration.pnpmOptions && rushConfiguration.pnpmOptions.environmentVariables) {
        configurationEnvironment = rushConfiguration.pnpmOptions.environmentVariables;
      }
    } else if (rushConfiguration.packageManager === 'yarn') {
      if (rushConfiguration.yarnOptions && rushConfiguration.yarnOptions.environmentVariables) {
        configurationEnvironment = rushConfiguration.yarnOptions.environmentVariables;
      }
    }

    return InstallHelpers._mergeEnvironmentVariables(process.env, configurationEnvironment, options);
  }

  /**
   * If the "(p)npm-local" symlink hasn't been set up yet, this creates it, installing the
   * specified (P)npm version in the user's home directory if needed.
   */
  public static async ensureLocalPackageManagerAsync(
    rushConfiguration: RushConfiguration,
    rushGlobalFolder: RushGlobalFolder,
    maxInstallAttempts: number,
    restrictConsoleOutput?: boolean
  ): Promise<void> {
    let logIfConsoleOutputIsNotRestricted: (message?: string) => void;
    if (restrictConsoleOutput) {
      logIfConsoleOutputIsNotRestricted = () => {
        /* noop */
      };
    } else {
      logIfConsoleOutputIsNotRestricted = (message?: string) => {
        // eslint-disable-next-line no-console
        console.log(message);
      };
    }

    // Example: "C:\Users\YourName\.rush"
    const rushUserFolder: string = rushGlobalFolder.nodeSpecificPath;

    if (!FileSystem.exists(rushUserFolder)) {
      logIfConsoleOutputIsNotRestricted('Creating ' + rushUserFolder);
      FileSystem.ensureFolder(rushUserFolder);
    }

    const packageManager: PackageManagerName = rushConfiguration.packageManager;
    const packageManagerVersion: string = rushConfiguration.packageManagerToolVersion;

    const packageManagerAndVersion: string = `${packageManager}-${packageManagerVersion}`;
    // Example: "C:\Users\YourName\.rush\pnpm-1.2.3"
    const packageManagerToolFolder: string = path.join(rushUserFolder, packageManagerAndVersion);

    const packageManagerMarker: LastInstallFlag = new LastInstallFlag(packageManagerToolFolder, {
      node: process.versions.node
    });

    logIfConsoleOutputIsNotRestricted(`Trying to acquire lock for ${packageManagerAndVersion}`);

    const lock: LockFile = await LockFile.acquireAsync(rushUserFolder, packageManagerAndVersion);

    logIfConsoleOutputIsNotRestricted(`Acquired lock for ${packageManagerAndVersion}`);

    if (!(await packageManagerMarker.isValidAsync()) || lock.dirtyWhenAcquired) {
      logIfConsoleOutputIsNotRestricted(
        Colorize.bold(`Installing ${packageManager} version ${packageManagerVersion}\n`)
      );

      // note that this will remove the last-install flag from the directory
      await Utilities.installPackageInDirectoryAsync({
        directory: packageManagerToolFolder,
        packageName: packageManager,
        version: rushConfiguration.packageManagerToolVersion,
        tempPackageTitle: `${packageManager}-local-install`,
        maxInstallAttempts: maxInstallAttempts,
        // This is using a local configuration to install a package in a shared global location.
        // Generally that's a bad practice, but in this case if we can successfully install
        // the package at all, we can reasonably assume it's good for all the repositories.
        // In particular, we'll assume that two different NPM registries cannot have two
        // different implementations of the same version of the same package.
        // This was needed for: https://github.com/microsoft/rushstack/issues/691
        commonRushConfigFolder: rushConfiguration.commonRushConfigFolder
      });

      logIfConsoleOutputIsNotRestricted(
        `Successfully installed ${packageManager} version ${packageManagerVersion}`
      );
    } else {
      logIfConsoleOutputIsNotRestricted(
        `Found ${packageManager} version ${packageManagerVersion} in ${packageManagerToolFolder}`
      );
    }

    await packageManagerMarker.createAsync();

    // Example: "C:\MyRepo\common\temp"
    FileSystem.ensureFolder(rushConfiguration.commonTempFolder);

    // Example: "C:\MyRepo\common\temp\pnpm-local"
    const localPackageManagerToolFolder: string = path.join(
      rushConfiguration.commonTempFolder,
      `${packageManager}-local`
    );

    logIfConsoleOutputIsNotRestricted(`\nSymlinking "${localPackageManagerToolFolder}"`);
    logIfConsoleOutputIsNotRestricted(`  --> "${packageManagerToolFolder}"`);

    // We cannot use FileSystem.exists() to test the existence of a symlink, because it will
    // return false for broken symlinks.  There is no way to test without catching an exception.
    try {
      FileSystem.deleteFolder(localPackageManagerToolFolder);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== 'ENOENT') {
        throw error;
      }
    }

    FileSystem.createSymbolicLinkJunction({
      linkTargetPath: packageManagerToolFolder,
      newLinkPath: localPackageManagerToolFolder
    });

    lock.release();
  }

  // Helper for getPackageManagerEnvironment
  private static _mergeEnvironmentVariables(
    baseEnv: NodeJS.ProcessEnv,
    environmentVariables?: IConfigurationEnvironment,
    options: {
      debug?: boolean;
    } = {}
  ): NodeJS.ProcessEnv {
    const packageManagerEnv: NodeJS.ProcessEnv = baseEnv;

    if (environmentVariables) {
      // eslint-disable-next-line guard-for-in
      for (const envVar in environmentVariables) {
        let setEnvironmentVariable: boolean = true;
        // eslint-disable-next-line no-console
        console.log(`\nProcessing definition for environment variable: ${envVar}`);

        if (baseEnv.hasOwnProperty(envVar)) {
          setEnvironmentVariable = false;
          // eslint-disable-next-line no-console
          console.log(`Environment variable already defined:`);
          // eslint-disable-next-line no-console
          console.log(`  Name: ${envVar}`);
          // eslint-disable-next-line no-console
          console.log(`  Existing value: ${baseEnv[envVar]}`);
          // eslint-disable-next-line no-console
          console.log(
            `  Value set in ${RushConstants.rushJsonFilename}: ${environmentVariables[envVar].value}`
          );

          if (environmentVariables[envVar].override) {
            setEnvironmentVariable = true;
            // eslint-disable-next-line no-console
            console.log(
              `Overriding the environment variable with the value set in ${RushConstants.rushJsonFilename}.`
            );
          } else {
            // eslint-disable-next-line no-console
            console.log(Colorize.yellow(`WARNING: Not overriding the value of the environment variable.`));
          }
        }

        if (setEnvironmentVariable) {
          if (options.debug) {
            // eslint-disable-next-line no-console
            console.log(`Setting environment variable for package manager.`);
            // eslint-disable-next-line no-console
            console.log(`  Name: ${envVar}`);
            // eslint-disable-next-line no-console
            console.log(`  Value: ${environmentVariables[envVar].value}`);
          }
          packageManagerEnv[envVar] = environmentVariables[envVar].value;
        }
      }
    }

    return packageManagerEnv;
  }
}
