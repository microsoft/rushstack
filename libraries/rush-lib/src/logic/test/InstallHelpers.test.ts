// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { type IPackageJson, JsonFile } from '@rushstack/node-core-library';
import { StringBufferTerminalProvider, Terminal } from '@rushstack/terminal';
import { TestUtilities } from '@rushstack/heft-config-file';

import { InstallHelpers } from '../installManager/InstallHelpers';
import { RushConfiguration } from '../../api/RushConfiguration';

describe('InstallHelpers', () => {
  describe('generateCommonPackageJson', () => {
    const originalJsonFileSave = JsonFile.save;
    const mockJsonFileSave: jest.Mock = jest.fn();
    let terminal: Terminal;
    let terminalProvider: StringBufferTerminalProvider;

    beforeAll(() => {
      JsonFile.save = mockJsonFileSave;
    });

    beforeEach(() => {
      terminalProvider = new StringBufferTerminalProvider();
      terminal = new Terminal(terminalProvider);
    });

    afterEach(() => {
      expect({
        output: terminalProvider.getOutput({ normalizeSpecialCharacters: true }),
        verbose: terminalProvider.getVerbose({ normalizeSpecialCharacters: true }),
        error: terminalProvider.getDebugOutput({ normalizeSpecialCharacters: true }),
        warning: terminalProvider.getWarningOutput({ normalizeSpecialCharacters: true }),
        debug: terminalProvider.getDebugOutput({ normalizeSpecialCharacters: true })
      }).toMatchSnapshot('Terminal Output');
      mockJsonFileSave.mockClear();
    });

    afterAll(() => {
      JsonFile.save = originalJsonFileSave;
    });

    it('generates correct package json with pnpm configurations', () => {
      const RUSH_JSON_FILENAME: string = `${__dirname}/pnpmConfig/rush.json`;
      const rushConfiguration: RushConfiguration =
        RushConfiguration.loadFromConfigurationFile(RUSH_JSON_FILENAME);
      InstallHelpers.generateCommonPackageJson(
        rushConfiguration,
        rushConfiguration.defaultSubspace,
        undefined,
        terminal
      );
      const packageJson: IPackageJson = mockJsonFileSave.mock.calls[0][0];
      expect(TestUtilities.stripAnnotations(packageJson)).toEqual(
        expect.objectContaining({
          pnpm: {
            overrides: {
              foo: '^2.0.0', // <-- unsupportedPackageJsonSettings.pnpm.override.foo
              quux: 'npm:@myorg/quux@^1.0.0',
              'bar@^2.1.0': '3.0.0',
              'qar@1>zoo': '2'
            },
            packageExtensions: {
              'react-redux': {
                peerDependencies: {
                  'react-dom': '*'
                }
              }
            },
            neverBuiltDependencies: ['fsevents', 'level'],
            pnpmFutureFeature: true
          }
        })
      );
    });
  });
});
