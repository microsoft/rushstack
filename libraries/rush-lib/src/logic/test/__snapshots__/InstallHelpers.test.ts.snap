// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstallHelpers generateCommonPackageJson generates correct package json with pnpm configurations: Terminal Output 1`] = `
Object {
  "debug": "",
  "error": "",
  "output": "",
  "verbose": "",
  "warning": "",
}
`;
